### PR TITLE
LSTM Time Series Functionality 

### PR DESCRIPTION
**Pull request by andrewberryy**
What type of PR is this? (check all applicable)
🍕 Feature

Description
I changed a couple files in the analysis folder so that the data being displayed on the graphs would output to a file called data.csv and be saved in the the LSTM folder. The LSTM folder contains a python script that runs the LSTM functionality and saves a results graph on your local machine. 

Run:

$ /Congestion-control/LSTM python3 lstm_network_emulator.py --epochs "# of epochs" --predictions "# of predictions" 

Added tests?
 🙋 no, because I need help
Added to documentation?
 📜 README.md

### CODE REVIEW

**Comment by a8nguyen** [comment]
Looks good. Ill update the docker image! Thanks!

File: src/analysis/tunnel_graph.py [hunk 1]
```diff
@@ -271,6 +271,17 @@ def plot_throughput_graph(self):
...
+                #output data to csv file
+                x_values = self.ingress_t[flow_id]
+                y_values = self.ingress_tput[flow_id]
+
+                with open('./LSTM/data.csv', 'w') as f:
+                    fields = ["Time","Throughput"]
+                    writer = csv.writer(f)
+                    writer.writerow(fields)
+                    writer.writerows(zip(x_values,y_values))
+
```
Comment (a8nguyen) [inline src/analysis/tunnel_graph.py:284]: intersting. is there a reason why ingress data are recorded for analysis but not egress

**Review by a8nguyen — APPROVED** [review]
I don't think we use the pkl file at all since we let the user build their own in `lstm_network_emulator.py` script. maybe consider removing it. And does *.csv and *. png files needed for the codebase?

File: src/analysis/tunnel_graph.py (hunk 1 above)
Comment (HalimaMalik) [reply src/analysis/tunnel_graph.py:284]: @andrewberryy

**Comment by a8nguyen** [comment]
I have some thoughts on the model. Might just explain how the model is working unstably:
- The obvious that we've been talking about. The small data sample
- At a closer look at the data, I am not sure if it is the floating point precision problem with measuring throughput but 0.003232, 0.001552 and other values are repeated multiple times (in different runs as well). So not only we have small sample size, the numerical space we are working with is very small
- The way the model is running is that we built from scratch every time so every run it's feeding a new pattern. So the lack of diverse patterns might be an issue here
- It's interesting that tanh works for that one example you showed since it's zero centered. I am not sure if certain things work well for certain patterns. But obviously, we are making predictions, so we can't know cheat and peak at the pattern while testing...
These are just my theoretical suspicions.

**Comment by HalimaMalik** [comment]
@andrewberryy

**Review by HalimaMalik — COMMENTED** [review]
@andrewberryy , Kindly examine the data used for the LSTM

**Comment by andrewberryy** [comment]
I appreciate your guys input! We have discussed that our data is an issue for the LSTM for a long time, the small data sample is the root of most of these issues. For a while I have been trying to figure out why 60s has been the cutoff for emulation time (we left the limit at 120 in our current code base) and why when you specify emulation time over 60 it glitches out and doesn't capture all the information for certain protocols but, I haven't been successful in any of my findings. I think if we can figure this out, we will solve a majority of these issues as the model would get more information and see more diverse patterns. (Perhaps working with different metrics rather than throughput could also be a way of finding more diverse patterns) As for the activation function, I am not sure why tanh worked so well with that dataset but I experimented a lot with sigmoid as well and got the exact same results as tanh. (maybe other activation functions would work better but I read that these two work best for LSTMs)

**Pull request merged**